agent/eni/watcher/watcher_windows.go

```diff
@@ -0,0 +1,164 @@
+// +build windows
+
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package watcher
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/eni/iphelperwrapper"
+	"github.com/aws/amazon-ecs-agent/agent/eni/networkutils"
+
+	log "github.com/cihub/seelog"
+	"github.com/pkg/errors"
+
+	"github.com/aws/amazon-ecs-agent/agent/engine/dockerstate"
+	"github.com/aws/amazon-ecs-agent/agent/statechange"
+)
+
+const (
+	defaultVirtualAdapterPrefix = "vEthernet"
+)
+
+// ENIWatcher maintains the state of attached ENIs to the instance. It also has supporting elements to
+// maintain consistency and update intervals
+// Platform specific fields are included at the end
+type ENIWatcher struct {
+	ctx                  context.Context
+	cancel               context.CancelFunc
+	updateIntervalTicker *time.Ticker
+	agentState           dockerstate.TaskEngineState
+	eniChangeEvent       chan<- statechange.Event
+	primaryMAC           string
+	interfaceMonitor     iphelperwrapper.InterfaceMonitor
+	notifications        chan int
+	netutils             networkutils.NetworkUtils
+}
+
+// NewWindowsWatcher is used to return an instance of the ENIWatcher
+// This method starts the interface monitor as well.
+// If the interface monitor cannot be started then we return an error
+func NewWindowsWatcher(ctx context.Context, primaryMAC string, state dockerstate.TaskEngineState,
```

Comment:
instead of having two exported methods New and NewWindowsWatcher, it would be cleaner to refactor the New() method and reuse it. remove the udevwrap from argument and initialize it inside for linux, let New returns (*ENIWatcher, error), and put the implementation in a private method that is defined in linux and windows separately . This will allow the user of ENIWatcher to not have different code for different platform.

Reply:
udevwrap for Linux and iphelperwrap for Windows are injected from the calling function. This is done so as to inject mocks during testing. Therefore, we can probably use a setter method for the same and go ahead with a common method.